ls1intum/Artemis
# `Development`: Extract plagiarism detection service

### Checklist
#### General
#### Server
#### Changes affecting Programming Exercises

### Motivation and Context
Plagiarism detection logic is distributed in controllers for each exercise type resources. This logic has a common domain and could be stored in a dedicated class. The current state will also require a copy&paste approach in case on implementing new triggers for plagiarism detection, like continuous plagiarism control.

### Description
This PR introduces a common `PlagiarismDetectionService` which own all plagiarism checks related logic. This class is implemented using the code moved from `ModelingExerciseResource`, `TextExerciseResource` and `ProgrammingExercisePlagiarismResource` classes. 

There is a new `PlagarismDetectionConfig` class, which holds the configuration for plagiarism checks. This is the first step that enables this config to be stored on the server, which will be necessary for the continuous plagiarism control in the future.

### Steps for Testing
Prerequisites:
- 1 Instructor and 2 Students
- 1 programming, 1 modeling and 1 text exercise

1. Create very similar submission to each of 3 exercises from both user accounts.
2. Go to exercise manage page and click on _Check plagiarism_.
3. Run plagiarism detection wich default settings.
4. Verify that plagiarism results are shown.
5. Run plagiarism detection wich similarity threshold changed to 100%.
6. Verify that **no** plagiarism result is shown.

### Review Progress
#### Performance Review
#### Code Review
#### Manual Tests

### Test Coverage

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| PlagiarismDetectionService | 100% | ✅                           |
